bugs/gh-275.txt

```diff
@@ -0,0 +1,736 @@
+=== Begin SRU Template ===
+[Impact]
+On s390x legacy terminals the use of square brackets [] is not available, therefore
+that makes it harder for those terminal to pass cloudinit configuration through kernel cmdline.
+Therefore, we are now supporting %-escaped characters, allowing those terminal to pass more complex
+config through kernel cmdline.
+
+[Test Case]
+
+```
+#!/bin/sh
+set -x
+# Manually deploy on a lxc container
+
+CMD_ARGUMENT="cc: runcmd%3A%20%5B%20echo%20testkernelcmdline%20%5D end_cc"
+
+restart_cloudinit() {
+   name=$1
+
+   lxc exec $name -- DI_LOG=stderr /usr/lib/cloud-init/ds-identify --force
+   lxc exec $name -- cloud-init clean --logs
+   lxc exec $name -- cloud-init init --local
+   lxc exec $name -- cloud-init init > /dev/null
+   lxc exec $name -- cloud-init modules --mode=config
+   lxc exec $name -- cloud-init modules --mode=final
+   lxc exec $name -- cloud-init status --wait --long
+}
+
+reproduce() {
+   series=$1
+   name=test-$SERIES-reproduce
+
+   lxc delete $name --force 2> /dev/null
+   lxc init ubuntu-daily:$SERIES $name
+   lxc config set $name environment.DEBUG_PROC_CMDLINE="$CMD_ARGUMENT"
+   lxc start $name
+   lxc exec $name -- cloud-init status --wait --long
+   lxc exec $name -- python3 -c "import os; print(os.environ.get('DEBUG_PROC_CMDLINE'))"
+
+   restart_cloudinit $name
+
+   expected_output="SUCCESS"
+   actual_output=$(lxc exec $name -- cat /var/log/cloud-init-output.log | grep "testkernelcmdline" || echo "SUCCESS")
+
+   if [ "$expected_output" = "$actual_output" ]; then
+       echo "SUCCESS: $series did not execute runcmd provided by $CMD_ARGUMENT"
+   else
+       echo "FAILURE: $series execute runcmd provided by $CMD_ARGUMENT"
+   fi
+
+   expected_output="Skipping module named runcmd, no 'runcmd' key in configuration"
+   actual_output=$(lxc exec $name -- cat /var/log/cloud-init.log | grep -o "$expected_output" || echo FAILURE)
+
+   if [ "$expected_output" = "$actual_output" ]; then
+       echo "SUCCESS: $series did not find runcmd provided by $CMD_ARGUMENT"
+   else
+       echo "FAILURE: $series has found runcmd provided by $CMD_ARGUMENT"
+   fi
+}
+
+verify() {
+   series=$1
+   ref=$SERIES-proposed
+   name=test-$SERIES-proposed
+
+   lxc delete $name --force 2> /dev/null
+   lxc-proposed-snapshot --proposed --upgrade cloud-init --publish $SERIES $ref
+   lxc init $ref $name
+   lxc config set $name environment.DEBUG_PROC_CMDLINE="$CMD_ARGUMENT"
```

Comment:
Nice work @lucasmoura TIL (today I learned) about setting environment variables in lxc config.